src/parcsr_mv/par_vector.c

```diff
@@ -564,6 +606,92 @@ hypre_ParVectorInnerProd( hypre_ParVector *x,
    return result;
 }
 
+/*--------------------------------------------------------------------------
+ * hypre_ParVectorInnerProdTagged
+ *
+ * Computes the "marked" inner product of two vectors x and y, i.e., an array
+ * of inner products computed from entries marked with tags.
+ *
+ * Resulting array is stored in host memory.
+ *  - iprod[0]: inner product of full vector
+ *  - iprod[i + 1]: inner product computed from entries marked with "i" tag
+ *
+ * Output data:
+ *  - num_tags_ptr: pointer to number of tags
+ *  - iprod_ptr: pointer to array of inner products
+ *               (allocated by the caller, otherwise allocated here)
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_ParVectorInnerProdTagged( hypre_ParVector  *x,
+                                hypre_ParVector  *y,
+                                HYPRE_Int        *num_tags_ptr,
+                                HYPRE_Complex   **iprod_ptr )
+{
+   MPI_Comm       comm        = hypre_ParVectorComm(x);
+   hypre_Vector  *x_local     = hypre_ParVectorLocalVector(x);
+   hypre_Vector  *y_local     = hypre_ParVectorLocalVector(y);
+   HYPRE_Int     *tags        = hypre_ParVectorTags(x);
+   HYPRE_Int      num_tags    = hypre_ParVectorNumTags(x);
+
+   HYPRE_Complex *iprod;
+   HYPRE_Complex *iprod_local;
+   HYPRE_Int      i;
+
+   /* Allocate output inner product array if needed */
+   if (*iprod_ptr == NULL)
+   {
+      iprod = hypre_CTAlloc(HYPRE_Complex, num_tags + 1, HYPRE_MEMORY_HOST);
+   }
+   else
+   {
+      iprod = *iprod_ptr;
+   }
+
+   /* Fallback to full vector inner product if num_tags == 1 or tags is NULL */
+   if (num_tags == 1 || !tags)
+   {
+      iprod[0] = hypre_ParVectorInnerProd(x, y);
+
+      *num_tags_ptr = 1;
+      *iprod_ptr = iprod;
+
+      return hypre_error_flag;
+   }
+
+   /* Initialize work array */
+   iprod_local = hypre_CTAlloc(HYPRE_Complex, num_tags + 1, HYPRE_MEMORY_HOST);
+
+   /* Compute local inner products */
+   hypre_SeqVectorInnerProdTagged(x_local, y_local, iprod_local);
+
+   /* Exit early in case of issues in the previous call */
+   if (hypre_error_flag)
+   {
+      return hypre_error_flag;
+   }
+
+   /* Global inner products */
+   hypre_MPI_Allreduce(iprod_local, &iprod[1], num_tags,
+                       HYPRE_MPI_COMPLEX, hypre_MPI_SUM, comm);
+
+   /* Compute total inner product */
+   iprod[0] = 0.0;
+   for (i = 0; i < num_tags; i++)
+   {
+      iprod[0] += iprod[i + 1];
+   }
+
+   /* Return results */
+   *num_tags_ptr = num_tags;
```

Comment:
Why is this needed if parVector has a num_tags member? Can we not query the num_tags directly from the calling function instead of retrieving it out here as an output from this function?

Reply:
In general yes, however for the krylov solvers that's trick because they don't know the vector type, i.e., it could be ParVector, StructVector, SStructVector... So I decided to return this as an output parameter in the innerprodtagged function